src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -1352,15 +1352,18 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
         void* ret = resource->allocate(num_bytes, stream);
         post_alloc_success(tid, likely_spill);
         return ret;
-      } catch (const std::bad_alloc& e) {
+      } catch (const rmm::out_of_memory& e) {
+        // rmm::out_of_memory is what is thrown when an allocation failed
+        // but there are other rmm::bad_alloc exceptions that could be
+        // thrown as well, which are handled by the std::exception case.
```

Comment:
should the other exceptions handled in the std::exception catch also conditionaly throw if `post_alloc_failed` returns false?

Reply:
The catch block for `std::exception` will always rethrow the original exception it caught.  Why would we want to hide the original exception if `post_alloc_failed` didn't return true? The boolean returned by `post_alloc_failed` indicates if we should retry the allocation or not. We don't want to retry an allocation unless the the first allocation failed for a reason we know we can potentially recover from by retrying.  That is the bug we are trying to fix with this patch. We saw an exception from RMM and misread it to be an allocation we should retry instead of a fatal error that we should not try to handle.  Also if it is not an OOM exception `post_alloc_failed` just moves the state of the thread back to RUNNING so it could recover if it needed to.